### PR TITLE
Feature/user authorization

### PR DESCRIPTION
**Pull request by open-schnick**
Checkout the feature files. Unit tests are still to come.

### CODE REVIEW

**Review by qvalentin — COMMENTED** [review]
I haven't checked everything

File: src/test/resources/UserAuthorization.feature [hunk 1]
```diff
@@ -1,44 +1,50 @@
...
+  When user requests accessToken with refreshToken "token" and userId 1234
+  Then response contains key "userId" and value "1234"
+  And response contains valid accessToken for user 1234
+  And response status code is 200
+
+Scenario: Successful request of existing accessToken with refreshToken.
+  Given accessToken with value "6bb9cb4f-7b51-4c0a-8013-ed7a34e56282" exists for user 1234
+  When user requests accessToken with refreshToken "token" and userId 1234
+  Then response contains key "userId" and value "1234"
+  And response contains valid accessToken for user 1234
```
[the remark this answers is not in the document]
Comment (open-schnick) [reply src/test/resources/UserAuthorization.feature:31]: no but there could be a usecase where the token is not valid anymore -> a new one will be returned.
Feel free to write one

[the remark this answers is not in the document]
Comment (open-schnick) [reply src/test/resources/UserAuthorization.feature:31]: i implemented a scenario

File: src/main/java/de/filefighter/rest/configuration/PrepareDataBase.java [hunk 2]
```diff
@@ -1,21 +1,52 @@
...
 
+    @Bean
+    CommandLineRunner veryImportantFileFighterStartScript() {
+        return args -> {
+            System.out.println();
+            System.out.println("-------------------------------< REST API >-------------------------------");
+            System.out.println();
+            System.out.println("  _____   _   _          _____   _           _       _                 ");
+            System.out.println(" |  ___| (_) | |   ___  |  ___| (_)   __ _  | |__   | |_    ___   _ __ ");
+            System.out.println(" | |_    | | | |  / _ \\ | |_    | |  / _  | | '_ \\  | __|  / _ \\ | '__|");
```
Comment (qvalentin) [inline src/main/java/de/filefighter/rest/configuration/PrepareDataBase.java:35]: looks ok

File: src/main/java/de/filefighter/rest/domain/user/business/UserBusinessService.java [hunk 3]
```diff
@@ -1,17 +1,108 @@
...
+    public long getUserCount() {
         return userRepository.count();
     }
+
+    public User getUserByUsernameAndPassword(String base64encodedUserAndPassword) {
+        if (!stringIsValid(base64encodedUserAndPassword))
+            throw new UserNotAuthenticatedException("Header was empty.");
+
+        //TODO: maybe filter unsupported characters?
+        if (!base64encodedUserAndPassword.matches("^"+AUTHORIZATION_BASIC_PREFIX+"(.*)$"))
```
Comment (qvalentin) [inline src/main/java/de/filefighter/rest/domain/user/business/UserBusinessService.java:41]: braun would be proud

Comment (open-schnick) [reply src/main/java/de/filefighter/rest/domain/user/business/UserBusinessService.java:41]: actually i changed them to be even better. just didnt push

**Comment by open-schnick** [comment]
@qvalentin check the userAuth feature again pls. Look at the status codes and response messages.

File: src/test/resources/UserAuthorization.feature [hunk 4]
```diff
@@ -1,44 +1,59 @@
...
+
+Scenario: Successful retrieval of existing accessToken with refreshToken.
+  Given accessToken with value "6bb9cb4f-7b51-4c0a-8013-ed7a34e56282" exists for user 1234
+  When user requests accessToken with refreshToken "token" and userId 1234
+  Then response contains key "userId" and value "1234"
+  And response contains valid accessToken for user 1234
+  And response status code is 200
+
+  # Better scenario description?
+Scenario: Successful retrieval of overwritten accessToken with refreshToken
```
Comment (qvalentin) [inline src/test/resources/UserAuthorization.feature:35]: of refreshed accessToken?

File: src/test/resources/SystemHealth.feature [hunk 5]
```diff
@@ -7,12 +7,14 @@ Background:
...
   Then response contains key "userCount" and value "0"
   And response contains key "uptimeInSeconds" and value of at least 1
 
 Scenario: SystemHealth is requested with users in db
   Given user 1234 exists
   And user 3214 exists
+  And the user waits for 2 second(s)
   When the systemHealth endpoint is requested
   Then response contains key "userCount" and value "2"
   And response contains key "uptimeInSeconds" and value of at least 1
```
Comment (qvalentin) [inline src/test/resources/SystemHealth.feature:20]: Status Keys?

File: src/test/resources/SystemHealth.feature [hunk 6]
```diff
@@ -7,12 +7,14 @@ Background:
```
Comment (qvalentin) [inline src/test/resources/SystemHealth.feature:7]: why is there not authentification?
Should every user be able to use this endpoint or only admins?

File: src/test/resources/UserAuthorization.feature (hunk 4 above)
Comment (open-schnick) [reply src/test/resources/UserAuthorization.feature:35]: its not refreshed.
The old one is deleted and a new one will be created.

File: src/test/resources/SystemHealth.feature (hunk 5 above)
Comment (open-schnick) [reply src/test/resources/SystemHealth.feature:20]: sure thing.

File: src/test/resources/SystemHealth.feature (hunk 6 above)
Comment (open-schnick) [reply src/test/resources/SystemHealth.feature:7]: Because its not needed in my opinion.

Comment (open-schnick) [reply src/test/resources/SystemHealth.feature:7]: The health endpoint doesnt really belong the the other rest end points.

**Pull request merged**